src/types/dict.rs

```diff
@@ -328,6 +334,25 @@ impl<'py> PyDictMethods<'py> for Bound<'py, PyDict> {
         BoundDictIterator::new(self.clone())
     }
 
+    #[cfg(Py_GIL_DISABLED)]
+    fn locked_for_each<F>(&self, closure: F) -> PyResult<()>
+    where
+        F: Fn(Bound<'py, PyAny>, Bound<'py, PyAny>) -> PyResult<()>,
+    {
+        let mut section = unsafe { std::mem::zeroed() };
+        unsafe { ffi::PyCriticalSection_Begin(&mut section, self.as_ptr()) };
+
+        for (key, value) in self {
+            if let Err(err) = closure(key, value) {
+                unsafe { ffi::PyCriticalSection_End(&mut section) };
```

Comment:
It occurred to me that this is probably not safe when the `closure` panics. I have to create a Guard here and call `PyCriticalSection_End` when dropping it after the iteration.

Reply:
For what it's worth, I have a half-completed wrapper around the critical section API using an RAII guard pattern. The tricky part is writing a test to make sure that the critical section is closed if a panic happens.

I'll probably have a PR for it tomorrow if you don't mind waiting for that :)

Reply:
Great! I just came up with the following wrapper but love to see your version tomorrow and use that one instead.

```rust
#[inline]
#[cfg(Py_GIL_DISABLED)]
fn critical_section<R, F>(op: *mut ffi::PyObject, closure: F) -> R
where
    F: FnOnce() -> R,
{
    struct Guard(ffi::PyCriticalSection);

    impl Drop for Guard {
        fn drop(&mut self) {
            unsafe {
                ffi::PyCriticalSection_End(&mut self.0);
            }
        }
    }
    
    let _guard = unsafe {
        let mut section = std::mem::zeroed();
        ffi::PyCriticalSection_Begin(&mut section, op);
        Guard(section)
    };
    closure()
}
```

Reply:
Mine's called `with_critical_section` but it's basically the same thing, I'll do a PR with docs and tests tomorrow.

Reply:
Here's the API @davidhewitt and I worked out on Friday:

```
/// Executes a closure on an object with an active critical section.
///
/// Locks modifications from other threads while the closure is executing.
/// This is structurally equivalent to the use of the paired
/// Py_BEGIN_CRITICAL_SECTION and Py_END_CRITICAL_SECTION macros exposed
/// by the C API.
///
/// This function is a no-op on Python versions older than 3.13, where
/// the critical section API is not exposed by the C API.
pub fn with_critical_section<F, R>(object: &Bound<'_, PyAny>, f: F) -> R
where
    F: FnOnce() -> R,
```

I think taking a `Bound` is nicer for a user-facing API and we decided that this is a safe abstraction, since while it is possible to do unsafe stuff, you need to use `unsafe` to do it (e.g. accessing dict or list internals inside a critical section).